test/test_common/simulated_time_system_test.cc

```diff
@@ -62,6 +63,105 @@ TEST_F(SimulatedTimeSystemTest, AdvanceTimeAsync) {
+  timer3->enableTimer(std::chrono::milliseconds(2));
+
+  // Expect watcher calls to happen in order since timers have different times.
+  testing::InSequence s;
+  EXPECT_CALL(watcher1, ready());
+  EXPECT_CALL(watcher2, ready());
+  EXPECT_CALL(watcher3, ready());
+  advanceMsAndLoop(5);
+}
+
+// Alarms that are scheduled to execute and are cancelled do not trigger.
+TEST_F(SimulatedTimeSystemTest, TimerOrderAndDisableAlarm) {
+  EXPECT_EQ(start_monotonic_time_, time_system_.monotonicTime());
+  EXPECT_EQ(start_system_time_, time_system_.systemTime());
+
+  ReadyWatcher watcher3;
+  Event::TimerPtr timer3 = scheduler_->createTimer([&] { watcher3.ready(); }, dispatcher_);
+
+  ReadyWatcher watcher2;
+  Event::TimerPtr timer2 = scheduler_->createTimer([&] { watcher2.ready(); }, dispatcher_);
+
+  ReadyWatcher watcher1;
+  Event::TimerPtr timer1 = scheduler_->createTimer(
+      [&] {
+        timer2->disableTimer();
+        watcher1.ready();
+      },
+      dispatcher_);
+
+  timer1->enableTimer(std::chrono::milliseconds(0));
+  timer2->enableTimer(std::chrono::milliseconds(1));
+  timer3->enableTimer(std::chrono::milliseconds(2));
+
+  // Expect watcher calls to happen in order since timers have different times. watcher2 ready() is
+  // not called because timer2 was disabled as part of timer1's execution.
+  testing::InSequence s;
+  EXPECT_CALL(watcher1, ready());
+  EXPECT_CALL(watcher3, ready());
+  advanceMsAndLoop(5);
+}
+
+// Change the registration time for an alarm that is already activated and verify that execution is
+// delayed.
+TEST_F(SimulatedTimeSystemTest, TimerOrderAndRescheduleAlarm) {
+  EXPECT_EQ(start_monotonic_time_, time_system_.monotonicTime());
+  EXPECT_EQ(start_system_time_, time_system_.systemTime());
+
+  ReadyWatcher watcher4;
+  Event::TimerPtr timer4 = scheduler_->createTimer([&] { watcher4.ready(); }, dispatcher_);
+
+  ReadyWatcher watcher3;
+  Event::TimerPtr timer3 = scheduler_->createTimer([&] { watcher3.ready(); }, dispatcher_);
+
+  ReadyWatcher watcher2;
+  Event::TimerPtr timer2 = scheduler_->createTimer([&] { watcher2.ready(); }, dispatcher_);
+
+  ReadyWatcher watcher1;
+  Event::TimerPtr timer1 = scheduler_->createTimer(
+      [&] {
+        timer2->enableTimer(std::chrono::milliseconds(0));
+        timer3->enableTimer(std::chrono::milliseconds(1));
+        watcher1.ready();
+      },
+      dispatcher_);
+
+  timer1->enableTimer(std::chrono::milliseconds(0));
+  timer2->enableTimer(std::chrono::milliseconds(1));
+  timer3->enableTimer(std::chrono::milliseconds(2));
+  timer4->enableTimer(std::chrono::milliseconds(3));
+
+  // timer1 is expected to run first and reschedule timers 2 and 3. watcher4 should fire before
+  // watcher2 since watcher4's registration is unaffected. watcher2 runs in the same iteration
+  // because it is scheduled with zero delay. timer3 executes in a later iteration because it is
+  // enabled with a non-zero timeout.
+  testing::InSequence s;
```

Comment:
Due to the 5ms advance, timers 1, 4 and 2 run in that order at T+5ms.  Timer3 executes at T+10ms due to the second advance.  The original and adjusted schedule times are only visible based on their effects on ordering.

Some things are not as easy to capture, like how many loop iterations were required to execute all the timers.  I guess that a check/prepare hook could be added to the scheduler to record some information on every loop iteration.

I take it you feel strongly about appending to strings instead of using strict mocks.  I'll look into it sometime on Friday.

Reply:
Is it significant to test how many loop iterations are required to execute all timers? I would have thought that ordering and time-timer-fired would be enough?

I don't feel that strongly about using output_, but it IMO it would make for a terser test that -- when it fails -- does so in a more actionable way. So I'll leave it up to you. But the main value of this PR is it fixes a bug, so I wouldn't hold it on that.

Reply:
Switched to using output_ to track ordering behavior.

Upcoming changes will likely change how timers interact with the event loop, so capturing number of iterations by tracking prepare calls in the output vector seems helpful.